Create date selector bar in stats

planarvoid opened the pull request:
Fixes #9137, #9067
This PR adds a date selecting bar to the DWMY stats. This bar is attached to the toolbar and doesn't hide on scroll. The enabled dates that could be iterated through the block are the same ones as in the Overview block. 
As part of this PR I'm also replacing the selected date in the title of the overview block with a bigger value for the selected date/unit (for example if there are selected Views and the value for the selected date is 10, there will be `10 Views` in the title).
I've removed the `Date` block since it no longer makes sense.

To test:
* Go to Stats/Insights
* There is no Date selector visible
* Go to Stats/DWMY
* The date selector appears
* The title of the Overview block is now the selected value
* Click on the buttons in the Date selector
  * The Selected date changes and new data is loaded
  * The right limit value disables the next button
  * The left limit value disables the previous button

![ripple_effect](https://user-images.githubusercontent.com/1079756/52200943-563d1980-286a-11e9-80d1-bd742a378c51.gif)

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

planarvoid committed:
Add date selector toolbar to stats

planarvoid committed:
Fix date selection from bars

planarvoid committed:
Fix selection of dates outside of the chart

planarvoid committed:
Replace loading block in Overview usecase

planarvoid committed:
Tests for date selection

planarvoid committed:
Add Traffic title to the overview block

planarvoid committed:
Revert "Add Traffic title to the overview block"

This reverts commit e51d8e8f9fea307c2619b42f196a8a19d62f349a.

planarvoid committed:
Fix ktlint issue

planarvoid committed:
Ellipsize date in the date selector when it's too long

planarvoid committed:
Finalize the Value designs

planarvoid committed:
Remove the date block from the DWMY screens

SylvesterWilmott commented:
Thanks for the changes @planarvoid ! Looks great

SylvesterWilmott commented:
@planarvoid one small thing I thought of.

Should we have some touch feedback on those chevrons? The typical circular ripple for icons

planarvoid committed:
Add ripple background to the date selector buttons

planarvoid commented:
@SylvesterWilmott I've added a circular ripple to the chevrons and updated the video (It's not always recorded for some reason but I think that's related to the Android emulator).

planarvoid committed:
Remove unused style

planarvoid commented:
@theck13 I've used this `android:background="?selectableItemBackgroundBorderless"` for the ripple effect. What would you recommend? One other thing I wasn't 100% sure about is the selectable chevrons. I created two vector files with different colors and a selector that switches between them. The other option would be to create a selector just for the colors and use this selector as the color of the vector drawable. Which one do you think we should use? thanks! 🙇

planarvoid committed:
Merge branch 'develop' into feature/create_date_selector_bar_in_stats

planarvoid committed:
Fixes to the icons and animation

planarvoid commented:
@theck13 thanks a lot for such a detailed feedback! it was super helpful. I've implemented the color selector and the icons the way you suggested (and removed the ones added in the PR). I didn't know I can just set `android:tint` instead of creating an icon. It might make sense to go through the code and locate the other places where I did the same thing. 
For the animation: I've tried many variations to find out what's causing this behaviour and I don't think it's just that we update the icons. It seems that updating/refreshing the UI stops the animation and that's something we cannot remove (since it's about the recycler view under the toolbar). In the end I added a 300ms delay on the next/previous click which gives the UI time to animate the ripple and still isn't long enough for the user to notice. It's not the best solution but I think it's good enough at the moment. Let me know what you think!

planarvoid committed:
Release notes update

theck13 commented:
> I've implemented the color selector and the icons the way you suggested (and removed the ones added in the PR).

Using the color selector looks good, @planarvoid!  The colors used in `grey_dark_grey_selector` for ***Stats*** are a little different than those used in `grey_dark_grey_lighten_10_selector` for ***History***.  We should probably use the same colors and selector file in both features for consistency.  @SylvesterWilmott, which version should we use for the enabled and disabled states of chevrons with a white background?

Version | Enabled | Disabled
-- | -- | --
1 | `grey_dark` `#2e4453` | `grey` `#87a6bc`
2 | `grey_dark` `#2e4453` | `grey_lighten_10` `#a8bece`

> It might make sense to go through the code and locate the other places where I did the same thing.

I've actually started consolidating all our drawables to a single white variation and adding tint to those drawable usages where possible.  We've merged two pull requests of many so far.  Don't worry about retroactively changing those for now.  We'll catch them eventually.

> In the end I added a 300ms delay on the next/previous click which gives the UI time to animate the ripple and still isn't long enough for the user to notice. It's not the best solution but I think it's good enough at the moment.

I'm not a fan of the delaying the interface even if it's 300ms since the delay affects updating the entire content.  If we could delay updating the chevrons only, that would be a decent workaround though it still doesn't seem like the best solution.

I spent some time trying to find where the interface update is causing the ripple to stop, but I didn't have any luck.  Even commenting out the entire [code snippet updating the toolbar views](https://github.com/wordpress-mobile/WordPress-Android/blob/218578145f771f33cb806189923b55054d644e09/WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsFragment.kt#L168-L182) didn't allow the ripple to complete before the interface refresh stopped it.  We update the enabled/disabled state of the chevron _immediately_ in the ***History*** feature, but the ripple doesn't stop.  Those two points make me think the problem must be due to the architecture somewhere.  If we intend to use this architecture throughout the app with other features, we should figure out a true fix now so we know how to address the issue in the future.

0nko commented on WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsViewModel.kt:
```diff
@@ -135,8 +167,66 @@ class StatsViewModel
+        val updatedDate = getDateLabelForSection(statsSection)
+        val currentState = showDateSelector.value
+        val statsGranularity = statsSection.toStatsGranularity()
+        if ((!shouldShowDateSelection && currentState?.isVisible != false) || statsGranularity == null) {
```
Perhaps it would be possible to simplify the update logic if we created a subclass of `LiveData` that would only get triggered if the new value is different. We could then just fire and forget without having to check for the current value.

planarvoid replied:
this is a good idea. It would be very useful to have this all over the place. What do you think about doing it in another PR? I think there is rarely a use case where we want to reemit the same item as before.

0nko replied:
That makes sense 👍.

planarvoid replied:
I've actually created a a Distinct LiveData in this PR but in this case it doesn't make sense because we only emit the section that has changed (for example `week`) every time we change the date - if we have a distinct call here, it stops working.

0nko reviewed (commented):
The selector looks great! 💯 While playing around with it I noticed that once you get to the very last column on the left, the left selector is disabled. I was just thinking that a user might want to see an earlier day, but perhaps it's intentional like this?

Another observation - when you move around using the date selector, the current chart bar highlight is delayed until all the data is refresh. I understand the wait for the view data that has to be downloaded but the chart bar highlight could be updated immediately, couldn't it? I feel like it would be more indicative of what data is actually being loaded, especially if you tap the arrow buttons a few times. It's just an idea. :)

I left a couple of comments in the code, too.

planarvoid commented:
@0nko thanks for the review! The previous item at the beginning of the graph is disabled on purpose. The same behaviour is implemented at the end of the graph. It could be a future improvement to move the items in the bar chart if we select the first column (or maybe move it even earlier). 
I'll look into refreshing the selected bar earlier. It's a little tricky :-)

SylvesterWilmott commented:
@theck13 thanks for your comments

>which version should we use for the enabled and disabled states of chevrons with a white background?

2 provides better contrast between enabled and disabled and as you mentioned is consistent with other parts of the app. @planarvoid let's just change the colour of the disabled chevron to `grey_lighten_10`

planarvoid committed:
Fix animation on next/previous date selected without delays

planarvoid commented:
@theck13 after loads of investigation I've found 2 issues and I actually managed to fix the problem without the delays. The first issue is that I was using a LinearLayout with weights for the date selector. Whenever we were setting the text value, the whole view was redrawn and the animation was stopped. I've extracted the layout and changed it to the constraint layout where everything is set and there is no need for redrawing. 
The second issue was the `isRefreshing` on the `SwipeToRefresh` layout. This seems to cause redrawing of the Toolbar. I've removed this state (since we should use the loading animation of the items anyway) and the animation works now. 
Let me know if it looks good to you now! I've also included some optimisation to not update the UI if there was no data change.

planarvoid committed:
Reuse color selector from history

planarvoid committed:
fix lint issue

planarvoid committed:
Fix ktlint issue

0nko commented on WordPress/src/main/java/org/wordpress/android/util/DistinctMutableLiveData.kt:
```diff
@@ -0,0 +1,21 @@
+
+import android.arch.lifecycle.MutableLiveData
+
+class DistinctMutableLiveData<T>(private val defaultValue: T) : MutableLiveData<T>() {
```
💯

theck13 reviewed (approved):
Great work finding the animation issue!  The ripple effect looks good and the delay has been removed.
